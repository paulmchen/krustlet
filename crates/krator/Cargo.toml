[package]
name = "krator"
version = "0.2.0"
authors = [
    "Matt Butcher <matt.butcher@microsoft.com>",
    "Matthew Fisher <matt.fisher@microsoft.com>",
    "Radu Matei <radu.matei@microsoft.com>",
    "Taylor Thomas <taylor.thomas@microsoft.com>",
    "Brian Ketelsen <Brian.Ketelsen@microsoft.com>",
    "Brian Hardock <Brian.Hardock@microsoft.com>",
    "Ryan Levick <rylevick@microsoft.com>",
    "Kevin Flansburg <kevin.flansburg@gmail.com>",
]
edition = "2018"
license-file = "../../LICENSE"
description = "A Kubernetes operator implementation in Rust"
repository = "https://github.com/deislabs/krustlet"
readme = "README.md"
keywords = [
    "state_machine",
    "operator",
    "kubernetes",
]

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["kube-native-tls"]
kube-native-tls = ["kube/native-tls", "kube-runtime/native-tls"]
rustls-tls = ["kube/rustls-tls", "kube-runtime/rustls-tls"]
derive = ["krator-derive"]
admission-webhook = ["warp", "json-patch"]

[dependencies]
async-trait = "0.1"
anyhow = "1.0"
tokio  = { version = "1.0", features = ["fs", "macros", "signal"] }
tokio-stream = { version = "0.1", features = ['sync'] }
k8s-openapi = { version = "0.11", default-features = false, features = ["v1_18"] }
kube = { version = "0.48", default-features = false }
kube-runtime = { version= "0.48", default-features = false }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
futures = { version = "0.3", default-features = false }
krator-derive = { version = "0.1", path = "../krator-derive", optional = true }
warp = { version = "0.3", optional = true, features = ["tls"] }
json-patch = { version = "0.2", optional = true }
tracing = { version = "0.1", features = ['log'] }
tracing-futures = "0.2"

[dev-dependencies]
kube-derive = "0.48"
schemars = "0.8.0"
serde_yaml = "0.8"
chrono = "0.4"
rand = "0.8"
tracing-subscriber = "0.2"
tokio  = { version = "1.0", features = ["fs", "macros", "signal", "rt-multi-thread"] }
opentelemetry-jaeger = "0.11"
tracing-opentelemetry = "0.11"
structopt="0.3"

[[example]]
name = "moose"
required-features = ["derive"]
