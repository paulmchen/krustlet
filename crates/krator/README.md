# Krator

**K**ubernetes **R**ust st**AT**e machine operat**OR**

[![krator documentation](https://docs.rs/krator/badge.svg)](https://docs.rs/krator)

Rust library for implementing your own operator in Rust.
