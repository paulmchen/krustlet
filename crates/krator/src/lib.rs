//! Exposes an API for creating state-machine-based Kubernetes Operators.

#![deny(missing_docs)]

mod manifest;
mod object;
mod operator;
mod runtime;

#[cfg(feature = "admission-webhook")]
pub mod admission;

pub mod state;

pub use manifest::Manifest;
pub use object::{ObjectState, ObjectStatus};
pub use operator::Operator;
pub use runtime::OperatorRuntime;
pub use state::{SharedState, State, Transition, TransitionTo};

#[cfg(feature = "derive")]
#[allow(unused_imports)]
#[macro_use]
extern crate krator_derive;

#[cfg(feature = "derive")]
#[doc(hidden)]
pub use krator_derive::*;
