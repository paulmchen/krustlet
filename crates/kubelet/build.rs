fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("cargo:rerun-if-changed=proto/pluginregistration/v1/pluginregistration.proto");

    let builder = tonic_build::configure()
        .format(true)
        .build_client(true)
        .build_server(true);

    // #[cfg(test)]
    // let builder = builder.build_server(true);
    // #[cfg(not(test))]
    // let builder = builder.build_server(false);

    builder.compile(
        &["proto/pluginregistration/v1/pluginregistration.proto"],
        &["proto/pluginregistration/v1"],
    )?;
    Ok(())
}
