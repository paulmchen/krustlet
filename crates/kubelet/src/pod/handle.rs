use std::collections::HashMap;

use tokio::io::{AsyncRead, AsyncSeek};
use tokio::sync::RwLock;
use tracing::{debug, error, info};

use crate::container::{
    ContainerKey, ContainerMapByName, Handle as ContainerHandle, HandleMap as ContainerHandleMap,
};
use crate::handle::StopHandler;
use crate::log::{HandleFactory, Sender};
use crate::pod::Pod;
use crate::provider::ProviderError;
use crate::volume::Ref;

/// Handle is the top level handle into managing a pod. It manages updating
/// statuses for the containers in the pod and can be used to stop the pod and
/// access logs
pub struct Handle<H, F> {
    container_handles: RwLock<ContainerHandleMap<H, F>>,
    pod: Pod,
    // Storage for the volume references so they don't get dropped until the runtime handle is
    // dropped
    // TODO: remove this; this is now part of the ModuleRunContext
    _volumes: HashMap<String, Ref>,
}

impl<H, F> std::fmt::Debug for Handle<H, F> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Handle")
            .field("pod", &self.pod.name())
            .finish()
    }
}

impl<H: StopHandler, F> Handle<H, F> {
    /// Creates a new pod handle that manages the given map of container names to
    /// [`ContainerHandle`]s. The given pod and client are used to maintain a reference to the
    /// kubernetes object and to be able to update the status of that object. The optional volumes
    /// parameter allows a caller to pass a map of volumes to keep reference to (so that they will
    /// be dropped along with the pod)
    pub fn new(
        container_handles: ContainerHandleMap<H, F>,
        pod: Pod,
        volumes: Option<HashMap<String, Ref>>,
    ) -> Self {
        Self {
            container_handles: RwLock::new(container_handles),
            pod,
            _volumes: volumes.unwrap_or_default(),
        }
    }

    /// Insert container `Handle` by `ContainerKey`.
    pub async fn insert_container_handle(&self, key: ContainerKey, value: ContainerHandle<H, F>) {
        let mut map = self.container_handles.write().await;
        map.insert(key, value);
    }

    /// Streams output from the specified container into the given sender.
    /// Optionally tails the output and/or continues to watch the file and stream changes.
    pub async fn output<R>(&self, container_name: &str, sender: Sender) -> anyhow::Result<()>
    where
        R: AsyncRead + AsyncSeek + Unpin + Send + 'static,
        F: HandleFactory<R>,
    {
        let mut handles = self.container_handles.write().await;
        let handle = handles
            .get_mut_by_name(container_name.to_owned())
            .ok_or_else(|| ProviderError::ContainerNotFound {
                pod_name: self.pod.name().to_owned(),
                container_name: container_name.to_owned(),
            })?;
        handle.output(sender).await
    }

    /// Signal the pod and all its running containers to stop and wait for them
    /// to complete.
    pub async fn stop(&self) -> anyhow::Result<()> {
        {
            let mut handles = self.container_handles.write().await;
            for (key, handle) in handles.iter_mut() {
                info!("Stopping container: {}", key);
                match handle.stop().await {
                    Ok(_) => debug!("Successfully stopped container {}", key),
                    // NOTE: I am not sure what recovery or retry steps should be
                    // done here, but we should definitely continue and try to stop
                    // the other containers
                    Err(e) => error!("Error while trying to stop pod {}: {:?}", key, e),
                }
            }
        }
        Ok(())
    }

    /// Wait for all containers in the pod to complete
    pub async fn wait(&mut self) -> anyhow::Result<()> {
        let mut handles = self.container_handles.write().await;
        for (_, handle) in handles.iter_mut() {
            handle.wait().await?;
        }
        Ok(())
    }
}

/// Generates a unique human readable key for storing a handle to a pod in a
/// hash. This is a convenience wrapper around [pod_key].
#[deprecated(
    since = "0.6.0",
    note = "Please use the new kubelet::pod::PodKey type. This function will be removed in 0.7"
)]
pub fn key_from_pod(pod: &Pod) -> String {
    #[allow(deprecated)]
    pod_key(pod.namespace(), pod.name())
}

/// Generates a unique human readable key for storing a handle to a pod if you
/// already have the namespace and pod name.
#[deprecated(
    since = "0.6.0",
    note = "Please use the new kubelet::pod::PodKey type. This function will be removed in 0.7"
)]
pub fn pod_key<N: AsRef<str>, T: AsRef<str>>(namespace: N, pod_name: T) -> String {
    format!("{}:{}", namespace.as_ref(), pod_name.as_ref())
}
