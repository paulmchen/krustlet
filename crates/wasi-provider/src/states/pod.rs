use crate::ModuleRunContext;
use crate::ProviderState;
use async_trait::async_trait;
use krator::{ObjectState, SharedState};
use kubelet::backoff::BackoffStrategy;
use kubelet::backoff::ExponentialBackoffStrategy;
use kubelet::pod::Pod;
use kubelet::pod::PodKey;
use kubelet::pod::Status;
use kubelet::state::common::{BackoffSequence, GenericPodState, ThresholdTrigger};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

pub(crate) mod completed;
pub(crate) mod initializing;
pub(crate) mod running;
pub(crate) mod starting;

/// State that is shared between pod state handlers.
pub struct PodState {
    key: PodKey,
    run_context: SharedState<ModuleRunContext>,
    errors: usize,
    image_pull_backoff_strategy: ExponentialBackoffStrategy,
    pub(crate) crash_loop_backoff_strategy: ExponentialBackoffStrategy,
}

#[async_trait]
impl ObjectState for PodState {
    type Manifest = Pod;
    type Status = Status;
    type SharedState = ProviderState;
    async fn async_drop(self, provider_state: &mut Self::SharedState) {
        {
            let mut handles = provider_state.handles.write().await;
            handles.remove(&self.key);
        }
    }
}

impl PodState {
    pub fn new(pod: &Pod) -> Self {
        let run_context = ModuleRunContext {
            modules: Default::default(),
            volumes: Default::default(),
        };
        let key = PodKey::from(pod);
        PodState {
            key,
            run_context: Arc::new(RwLock::new(run_context)),
            errors: 0,
            image_pull_backoff_strategy: ExponentialBackoffStrategy::default(),
            crash_loop_backoff_strategy: ExponentialBackoffStrategy::default(),
        }
    }
}

#[async_trait]
impl GenericPodState for PodState {
    async fn set_modules(&mut self, modules: HashMap<String, Vec<u8>>) {
        let mut run_context = self.run_context.write().await;
        run_context.modules = modules;
    }
    async fn set_volumes(&mut self, volumes: HashMap<String, kubelet::volume::Ref>) {
        let mut run_context = self.run_context.write().await;
        run_context.volumes = volumes;
    }
    async fn backoff(&mut self, sequence: BackoffSequence) {
        let backoff_strategy = match sequence {
            BackoffSequence::ImagePull => &mut self.image_pull_backoff_strategy,
            BackoffSequence::CrashLoop => &mut self.crash_loop_backoff_strategy,
        };
        backoff_strategy.wait().await;
    }
    async fn reset_backoff(&mut self, sequence: BackoffSequence) {
        let backoff_strategy = match sequence {
            BackoffSequence::ImagePull => &mut self.image_pull_backoff_strategy,
            BackoffSequence::CrashLoop => &mut self.crash_loop_backoff_strategy,
        };
        backoff_strategy.reset();
    }
    async fn record_error(&mut self) -> ThresholdTrigger {
        self.errors += 1;
        if self.errors > 3 {
            self.errors = 0;
            ThresholdTrigger::Triggered
        } else {
            ThresholdTrigger::Untriggered
        }
    }
}
