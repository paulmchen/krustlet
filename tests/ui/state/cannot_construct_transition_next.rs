// Test that State<T> can only transition to State<T>
// edition:2018
extern crate async_trait;
extern crate kubelet;
extern crate krator;

use krator::{Transition, state::StateHolder, ObjectState};
use kubelet::pod::{Pod, Status, state::Stub};

struct PodState;
struct ProviderState;

#[async_trait::async_trait]
impl ObjectState for PodState {
    type Manifest = Pod;
    type Status = Status;
    type SharedState = ProviderState;
    async fn async_drop(self, _provider_state: &mut ProviderState) { }
}

fn main() {
    // This fails because `state` is a private field. Use Transition::next classmethod instead.
    let _transition = Transition::<PodState>::Next(StateHolder {
        state: Box::new(Stub),
    });
}
