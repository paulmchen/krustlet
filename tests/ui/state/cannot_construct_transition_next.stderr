error[E0451]: field `state` of struct `StateHolder` is private
  --> $DIR/cannot_construct_transition_next.rs:24:9
   |
24 |         state: Box::new(Stub),
   |         ^^^^^^^^^^^^^^^^^^^^^ private field

error: aborting due to previous error

For more information about this error, try `rustc --explain E0451`.
